### PR TITLE
Add fillna for ListMethods.

### LINKED ISSUE #13274: [ENH] add `.fillna` method to `ListMethods`
**Describe the bug**
.fillna does not work with `ListColumns`.

**Steps/Code to reproduce bug**
```
x = cudf.Series([0, 1, None])
```
```
x.fillna(5)
0    0
1    1
2    5
dtype: int64
```
```
y = cudf.Series([[0, 1], [2, 3], [None, None]])
y.fillna(5) 
```
```
0    [0.0, 1.0]
1    [2.0, 3.0]
2    [nan, nan]
dtype: list
```

**Expected behavior**
`2    [nan, nan]` should be `2    [5, 5]`

**Environment overview (please complete the following information)**
cudf 23.06 installed via pip

**Environment details**
Not available.

**Additional context**

### PR DESCRIPTION
**Pull request by thomcom**
Closes #13274

## Description
This PR adds `list.fillna()` which performs `fillna` at every level of a `ListColumn`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/column/lists.py [hunk 1]
```diff
@@ -725,3 +725,57 @@ def astype(self, dtype):
...
+                method = "bfill"
+        result = self._column.copy(deep=True)
+        current = result
+        while len(current._base_children) != 0:
+            new_children = [child for child in current._base_children[:-1]]
+            new_children.append(
+                current._base_children[-1].fillna(value, method)
+            )
+            current.set_base_children(tuple(new_children))
+            current = current._base_children[-1]
```
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf/cudf/core/column/lists.py:780]: Unfortunately, this doesn't do the right thing for top-level nulls:
```
s = cudf.Series([None, [1]])
s.list.fillna([2])
```
Doesn't do anything AFAICT. This is because if there are top-level nulls, then the information is encoded in the null-mask (not as an entry in the array).

File: python/cudf/cudf/core/column/lists.py [hunk 2]
```diff
@@ -725,3 +725,57 @@ def astype(self, dtype):
...
+            if method == "pad":
+                method = "ffill"
+            elif method == "backfill":
+                method = "bfill"
+        result = self._column.copy(deep=True)
+        current = result
+        while len(current._base_children) != 0:
+            new_children = [child for child in current._base_children[:-1]]
+            new_children.append(
+                current._base_children[-1].fillna(value, method)
```
Comment (wence-) [inline python/cudf/cudf/core/column/lists.py:777]: If `base_children[-1]` is a `ListColumn` how does this work? AFAICT, `fillna` is implemented on all column types but raises a libcudf error if not supported (which it is not on `ListColumn`s).

**Review by wence- — CHANGES_REQUESTED** [review]
Am not sure that everything is as it seems?

File: python/cudf/cudf/core/column/lists.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cudf/cudf/core/column/lists.py:781]: Answering your question requires a fairly deep dive into `fillna` in both pandas and cudf.

In pandas, `pd.Series([None, [1]]).fillna([2])` is not supported as `value` must be either a scalar or a dict.

`pd.Series([None, [1]]).fillna({0: [2]})` is then supported, and is a combination of `replace` and `fillna` as it will only attempt to `fillna` the indices specified by the key. This seems like a terrible solution in cudf however as it would require something like 

```
xna = x.isna()
keys = xna.index[xna].values
values = [[2]] * len(keys)
nadict = dict(zip(keys, values))
x.fillna(nadict)
```

Which I don't think will ever be appropriate for cudf work.

Adding to the complexity of this implementation, pandas obviously supports lists with arbitrary dtypes so calls like

```
p = pd.Series([None, [1, None], [[1, None]]])
p.fillna(1)
```

are fine, but note that `fillna` only operates on the first level of the objects in the series.

cudf simply doesn't support mixed types in `ListColumn`, with varying errors thrown for each of

```
c = cudf.from_pandas(p)
c.fillna(cudf.Series([1]))
c.fillna([1])
c.fillna(1)
```

and no method for correcting them other than a deep dive into cython and even libcudf (which I'm not offering to do at this time. ;).

Therefor the solution I've proposed is to replace `None` at the leaf level of arbitrary `ListColumns` without violating or expanding the existing constraints on cudf's `fillna`. See the examples in the included tests. Note that if I created an initialized `Series` from any of these examples with `None` anywhere but the leaf level, cudf will allow it because arrow permits `None` in any dtype. However, attempting to replace these `None` will result in an error, with arrow-based dtype violations.

The right answer might be to not implement `.list.fillna` at all, and simply put a `GeoColumn` specific fillna in place for the issue in cuspatial that this PR is linked to. 

What do you think? Does my implementation cover a meaningful enough use case, or does it just muddy the pool further?

File: python/cudf/cudf/core/column/lists.py (hunk 2 above)
Comment (thomcom) [reply python/cudf/cudf/core/column/lists.py]: It doesn't raise an error, it does nothing when `base_children[-1]` is a `ListColumn` IIRC.

```
c = cudf.Series([[1, None], [2, None], [3, None]])
c._column.fillna(1)
```
No error, no change.

File: python/cudf/cudf/core/column/lists.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cudf/cudf/core/column/lists.py]: Oh I was mistaken that one can create a series like `cudf.Series([None, [1, None], [[2, None]]])`, one can only create empty series in this fashion: `cudf.Series([None, [None], [[None]]])`, but adding values at any level of mixed nesting can't be represented in arrow.

[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf/cudf/core/column/lists.py]: Thanks for the summary, let me just collect my thoughts:

tl;dr: I think this behaviour is fine, _but_:
1. can we explicitly raise an informational message in `ListColumn.fillna` (rather than the generic `RuntimeError`). 
1. can we then catch that error in `.list.fillna` and provide a useful informational error that one can only transform leaves in certain circumstances
1. add tests for these exceptional cases that they `with pytest.raises(WhateverErrorWeThoughtOf):`

## pandas behaviour

`.list.fillna` doesn't exist

`.fillna` On a "list" column (really object dtype) doesn't traverse to the leaves and only replaces top-level nulls

## current cudf behaviour

`.list.fillna` doesn't exist

`.fillna` on a list column either returns the column unchanged or (if there are top-level nulls) produces a `RuntimeError`

## this PR

`.fillna` is unchanged

`.list.fillna` exists. It leaves top-level nulls untouched and only transforms "leaf" nulls. If the leaf nulls are at the same level of nesting as the list dtype it succeeds, if any leaf null is at a different nesting level (e.g. `[[[1]], [None]]`) it produces a `RuntimeError` (because it's just calling `.fillna` on the column which is a `[[1], None]` and so we're in the old implementation realm).

AIUI, in geospatial you want this because you might read very structured list columns (point data) from files and therefore have a guarantee that any nulls that might appear will only ever be at the same level as everyone else. You want to be able to replace these with some actual number.

I think this is a reasonable semantic for `.list.fillna`. The only other one I can think of that might be reasonable would be to say `.list.fillna` operates at all levels and wraps the scalar value in as many singleton lists as necessary to make the nesting match (so if we had `[[[1, 2], None], None, [None], [[None]]]` and did `.list.fillna(3)` we would get `[[[1, 2], [3]], [[3]], [[3]]]`). That may be too magic though.

As noted above, I think the additional part necessary here is to provide nicer failure modes for the unsupported behaviour.

**Comment by thomcom** [comment]
Closing this due to cuspatial being its driving necessity.
